src/psyclone/psyir/nodes/omp_directives.py

```diff
@@ -638,6 +705,946 @@ def _encloses_omp_directive(self):
+                    type(ref).__name__))
+            if (ref not in firstprivate_list and ref not in private_list and
+                ref not in shared_list):
+                firstprivate_list.append(ref.copy())
+
+        stop_val_refs = stop_val.walk(Reference)
+        for ref in stop_val_refs:
+            if isinstance(ref, (ArrayReference, ArrayOfStructuresReference)):
+                raise GenerationError("{0} not yet supported in the stop "
+                                      "variable of the primary Loop in a "
+                                      "OMPTaskDirective node.".format(
+                    type(ref).__name__))
+            if (ref not in firstprivate_list and ref not in private_list and
+                ref not in shared_list):
+                firstprivate_list.append(ref.copy())
+
+        
+        step_val_refs = step_val.walk(Reference)
+        for ref in step_val_refs:
+            if isinstance(ref, (ArrayReference, ArrayOfStructuresReference)):
+                raise GenerationError("{0} not yet supported in the step "
+                                      "variable of the primary Loop in a "
+                                      "OMPTaskDirective node.".format(
+                    type(ref).__name__))
+            if (ref not in firstprivate_list and ref not in private_list and
+                ref not in shared_list):
+                firstprivate_list.append(ref.copy())
+
+        # Finished handling the loop bounds now
+
+        # Recurse to the children
+        for child_node in node.children[3].children:
+            self._evaluate_node(child_node, private_list, firstprivate_list,
+                                shared_list, in_list, out_list)
+
+        # Remove any stuff added to proxy_loop_vars etc. if needed
+        if to_remove is not None:
+            self._proxy_loop_vars.pop(to_remove)
+
+    def _evaluate_ifblock(self, node, private_list, firstprivate_list,
+                          shared_list, in_list, out_list):
+        '''
+        TODO: docstring
+        '''
+        # TODO: Look at the ifblock itself first.
+        for ref in node.condition.walk(Reference):
+            self._evaluate_readonly_reference(ref, private_list,
+                                              firstprivate_list, shared_list,
+                                              in_list, out_list)
+
+        # Recurse to the children
+        # If block
+        for child_node in node.if_body.children:
+            self._evaluate_node(child_node, private_list, firstprivate_list,
+                                shared_list, in_list, out_list)
+        # Else block if present
+        if node.else_body is not None:
+           for child_node in node.else_body.children:
+               self._evaluate_node(child_node, private_list, firstprivate_list,
+                                   shared_list, in_list, out_list)
+
+    def _evaluate_node(self, node, private_list, firstprivate_list,
```

Comment:
Possibly, you could tidy this up by implementing a Visitor class that inherits from `psyclone.psyir.backend.PSyIRVisitor`. Your `_evaluate_assignment` would then be `assignment_node`, `loop_node` etc. etc. I say 'possibly' because I'm not totally certain that PSyIRVisitor will fully support what you're doing but it's definitely worth a look. For examples of implementations see e.g. `psyclone.psyir.backend.c` or `psyclone.psyad.adjoint_visitor`.